demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,680 @@
+   "metadata": {},
+   "source": [
+    "## Split edges\n",
+    "\n",
+    "We want to split the edges into parts that can be used for our link prediction task:\n",
+    "* the oldest edges are used to create the graph structure\n",
+    "* the recent edges are what we are interested in predicting - we randomly split this part further into training and test sets."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of edges in graph: 37929\n",
+      "Number of edges in training set: 9524\n",
+      "Number Of edges in test set: 3119\n"
+     ]
+    }
+   ],
+   "source": [
+    "# subset of edges to split\n",
+    "train_subset = 0.25\n",
+    "test_subset = 0.25\n",
+    "\n",
+    "# number of edges to be kept in the graph\n",
+    "num_edges_graph = int(len(edges) * (1 - train_subset))\n",
+    "\n",
+    "# keep older edges in graph, and predict more recent edges\n",
+    "edges_graph = edges[:num_edges_graph]\n",
+    "edges_other = edges[num_edges_graph:]\n",
+    "\n",
+    "# split recent edges further to train and test sets\n",
+    "test_mask = np.random.rand(len(edges_other)) < test_subset\n",
+    "edges_train = edges_other[~test_mask]\n",
+    "edges_test = edges_other[test_mask]\n",
+    "\n",
+    "print(\n",
+    "    f\"Number of edges in graph: {len(edges_graph)}\\n\"\n",
+    "    f\"Number of edges in training set: {len(edges_train)}\\n\"\n",
+    "    f\"Number Of edges in test set: {len(edges_test)}\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create a StellarGraph\n",
+    "\n",
+    "Now we can use the edges we've reserved for the graph to create an instance of the StellarGraph class"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "graph = StellarGraph(\n",
+    "    nodes=nodes,\n",
+    "    edges=edges_graph,\n",
+    "    source_column=\"source\",\n",
+    "    target_column=\"target\",\n",
+    "    edge_weight_column=\"time\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create link examples for training and testing\n",
+    "\n",
+    "The edges that we've reserved for the train/test sets become examples of positive links. \n",
+    "\n",
+    "We also need to randomly generate some examples of negative links in order to train our classifier."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_link_examples(g, edges):\n",
+    "    pos = [(src, tgt) for (src, tgt) in edges[[\"source\", \"target\"]].values]\n",
+    "    neg = sample_negative_examples(g, pos)\n",
+    "    return pos, neg\n",
+    "\n",
+    "\n",
+    "def sample_negative_examples(g, positive_examples):\n",
```

Comment:
I'm sure you realise this, but it's lucky this is a graph with only 151 edges, since this function is O(n^2) :smile:

Reply:
I guess an alternative would be to just do the random sampling in a while loop and each time check that it doesn't exist in the graph - do you think that would be better?

Looks like our `EdgeSplitter._sample_negative_examples_global` has a slightly different approach too, but I thought it was simple/clear enough to just do it in the notebook instead.

Reply:
In section 3 of the paper:

> The remaining 25% are considered as positive links and
we sample an equal number of negative edges randomly

is the only reference to how this is done. I would expect the scores to improve further if negative edges were forced to not overlap with any of the positives at all? So in this case I'm not sure it explains the improvement in our code.